### PR TITLE
Remove GHC 8.4

### PR DESCRIPTION
**Pull request by alex-mckenna**
To support using -XQuantifiedConstraints in the partial evaluator,
support for GHC 8.4 is dropped. Conditionally compiled code for
pre GHC 8.4 is removed, and version bounds are bumped to a minimum
of GHC 8.6.

### CODE REVIEW

**Comment by martijnbastiaan** [comment]
You can remove 8.4 from here too:

https://github.com/clash-lang/clash-compiler/blob/12fa89070d2cf55c133ab93af2faab2b611659f3/.ci/docker/build-and-publish-docker-image.sh#L10

File: clash-prelude/src/Clash/Sized/Internal/Unsigned.hs [hunk 1]
```diff
@@ -498,18 +496,21 @@ shiftL#, shiftR#, rotateL#, rotateR# :: forall n .KnownNat n => Unsigned n -> In
...
+       in U ((naturalShiftL v i') `mod` m)
 #else
       U ((shiftL v i) `mod` m)
 #endif
     else
       error ("'shiftL undefined for negative number: " ++ show i)
  where
 #if MIN_VERSION_base(4,15,0)
-  m = 1 `naturalShiftL` naturalToWord (natVal (Proxy @n))
+  sz = naturalToWord (natVal (Proxy @n))
```
Comment (martijnbastiaan) [inline clash-prelude/src/Clash/Sized/Internal/Unsigned.hs:509]: `natToNum @n`?

Comment (alex-mckenna) [reply clash-prelude/src/Clash/Sized/Internal/Unsigned.hs:509]: Eh, my feeling is that these internal modules could do with a little spring cleaning in places anyway. Leave it for another PR?

Comment (martijnbastiaan) [reply clash-prelude/src/Clash/Sized/Internal/Unsigned.hs:509]: Sure thing. I commented because you changed it, that's all :)

Comment (alex-mckenna) [reply clash-prelude/src/Clash/Sized/Internal/Unsigned.hs:509]: If this is approved, feel free to merge :wink:

**Review by martijnbastiaan — APPROVED** [review]
LGTM. @leonschoorl can you think of anything else?

**Comment by leonschoorl** [comment]
I found a couple more

File: STYLE.rst [hunk 2]
```diff
@@ -235,10 +235,6 @@ blank line. Pragmas should be ordered alphabetically. Example:
     {-# LANGUAGE FlexibleInstances #-}
     {-# LANGUAGE QuasiQuotes #-}
 
-    #if __GLASGOW_HASKELL__ < 806
-    {-# LANGUAGE TypeInType #-}
-    #endif
-
```
Comment (alex-mckenna) [inline STYLE.rst:241]: I didn't even think to check the style guide (said every clash developer in unison)

**Comment by leonschoorl** [comment]
Was surprised to see it still compile with the removal of `HsVersions.h`
Must be getting it from the host compiler, because it's still used in all newer version:
https://github.com/clash-lang/clash-compiler/blob/12fa89070d2cf55c133ab93af2faab2b611659f3/clash-ghc/src-bin-9.0/Clash/GHCi/UI/Monad.hs#L37

File: STYLE.rst (hunk 2 above)
Comment (leonschoorl) [reply STYLE.rst:241]: Me neither.
But my global-find did.

Comment (alex-mckenna) [reply STYLE.rst:241]: My ~only~ mistake was searching only `clash-{prelude,lib,ghc}`

**Comment by alex-mckenna** [comment]
> Was surprised to see it still compile with the removal of HsVersions.h

I was initially surprised when I noticed it was only in the 8.4 dir, then assumed that other people who did the GHC stuff for 8.6, 8.8, 8.10, 9.0 knew what they were doing (given that it still compiled, and therefore works :wink: )

File: clash-prelude/src/Clash/Sized/Internal/Unsigned.hs (hunk 1 above)
Comment (christiaanb) [reply clash-prelude/src/Clash/Sized/Internal/Unsigned.hs:509]: Please check that the generated Core is as efficient if you do decide to switch to `natToNum`.

**Pull request merged**